docs/design/unified_sidecar_and_process_monitor.md

```diff
@@ -0,0 +1,123 @@
+
+When the new launcher is used, both the main container and the sidecar container will run the same image, but the sidecar container will have a different set of arguments that cause it to only copy the fdbserver binary, where necessary, and to open up a listener. The main container will load its configuration file, launch the configured server processes, and open a listener. That listener will be an HTTP server that provides an API for checking the hashes of files and other local configuration, similar to the API the sidecar currently provides. However, the fdb-kubernetes-monitor API will be not cover all use cases of the sidecar API. It will provide the following API calls:
+
+* `get_current_configuration` - returns the current process configuration
+* `check_hash` - check the hash of a file inside its configuration directory
+* `load_configuration` - load the latest configuration from the configuration directory into memory.
+* `ready` - returns an `OK` response
+* `substitutions` - returns the values for any environment variables that are used in the process configuration
+
+When we are transitioning from the old launcher to the new launcher for a cluster, we will include both a monitor conf file and the kubernetes-monitor conf file in the config map. Once the transition is complete, we will remove the monitor conf file from the sidecar.
+
+This API will have an option to listen on a TLS connection and restrict the incoming connections using the same peer verification options we support in the current sidecar.
+
+### Process Configuration
+
+The process configuration will be represented in a JSON file that contains a ProcessConfiguration object. The spec for this file is describe below.
+
+	ProcessConfiguration {
+		version: string - The version of FoundationDB the process should run. This will determine the path to the fdbserver process.
+		arguments: []Argument - The arguments to the fdbserver process
+	}
+
+	Argument {
+		type: ArgumentType - Determines how the value is generated. The default is Literal
+		value: string - For Literal arguments, the value for the argument
+		values: []Argument - For Concatenate arguments, a list of smaller arguments that are combined to form a single argument
+		source: string - For Environment arguments, the name of an environment variable to use as the source of the argument.
+		multiplier: integer - For ProcessNumber arguments, a multiplier to apply to the process number
+		offset: integer - For ProcessNumber arguments, an offset to apply to the process number
+	}
+
+	ArgumentType enum {
+		Literal - A value provided in the argument
+		Concatenante - A combination of other arguments
+		Environment - A value from an environment variable
+		ProcessNumber - A value calculated from the process number. The processes are numbered starting from 1.
+	}
+
+Example configuration, in JSON form:
+
+```json
```

Comment:
As the goal/scope of this process is to spawn `fdbserver` processes only, not generic process spawning - did you consider a configuration model where only defined fields/arguments accepted? If so, what are the reasons to not go for it.

Reply:
I think that would run the risk of creating more brittleness and duplicate work if we need to add new arguments for new versions of FDB. For instance, if FDB 7.1 involves a new flag that users may want to use, we would need to update the sidecar to permit it. We would also need wildcard support for fields like `--locality_*`, and maybe `--knob_*` as well.